_episodes/38.md

```diff
@@ -0,0 +1,475 @@
+demonstrating how polymorphic table functions can work and what they can do for
+you. But that's a simple example, and now we can look at some which are a little
+more complex and a little more practical.
+
+#### CSVreader
+
+If you've ever tried to create a table from a CSV file, you know it can be a
+painful experience. It has to be very explicit, very diligent, and there's a lot
+of manual cross-checking involved in ensuring that each column aligns perfectly
+and is correctly typed for the columns present in the CSV. Enter polymorphic
+table functions, here to save the day.
+
+The table function itself will be defined on the connector, so all the end user
+needs to worry about is its signature:
+
+```sql
+FUNCTION CSVreader (
+    Filename VARCHAR(1000),
+    FloatCols DESCRIPTOR DEFAULT NULL,
+    DateCols DESCRIPTOR DEFAULT NULL)
+RETURNS TABLE
+NOT DETERMINISTIC
+CONTAINS SQL
+```
+
+One key thing to note here is the `DESCRIPTOR` type. It is a type that describes
+a list of column names, and there will be a function to convert a parameterized
+list to the `DESCRIPTOR` type. Other than that, everything else here does what
+you'd expect - you pass the function the name of the CSV file, the columns which
+should be typed as floats, and the columns which should have a date typing. All
+unspecified columns will still be handled as `varchar`. Calling the function
+might look something like:
+
+```sql
+SELECT
+    *
+FROM
+    TABLE(
+        CSVreader(
+            Filename => 'my_file.csv',
+            FloatCols => DESCRIPTOR("principle", "interest")
+            DateCols => DESCRIPTOR("due_date")
+        )
+    );
+```
+
+Given a CSV with this content:
+
+```csv
+docno,name,due_date,principle,interest
+123,Mary,01/01/2014,234.56,345.67
+234,Edgar,01/01/2014,654.32,543.21
+```
+
+Such a function would return a table that looks like:
+
+| docno | name  | due_date   | principle | interest  |
+|-------|-------|------------|-----------|-----------|
+| 123   | Alice | 2014-01-01 | 234.56    | 345.67    |
+| 234   | Bob   | 2014-01-01 | 654.32    | 543.21    |
+
+With a well-written PTF, the days of toiling over parsing a CSV into SQL are
+over!
+
+#### Pivot
+
+Pivot is an oft-requested feature which hasn't been built in Trino because it
+isn't a part of the standard SQL specification. A `PIVOT` keyword or built-in
+function isn't planned, but with PTFs, we can support `PIVOT`-like functionality
+without needing to deviate from SQL.
+
+A `PIVOT` PTF might have the following definition:
+
+```sql
+FUNCTION Pivot (
+    Input_table TABLE PASS THROUGH WITH ROW SEMANTICS,
+    Output_pivot_columns DESCRIPTOR,
+    Input_pivot_columns1 DESCRIPTOR,
+    Input_pivot_columns2 DESCRIPTOR DEFAULT NULL,
+    Input_pivot_columns3 DESCRIPTOR DEFAULT NULL,
+    Input_pivot_columns4 DESCRIPTOR DEFAULT NULL,
+    Input_pivot_columns5 DESCRIPTOR DEFAULT NULL
+)
+RETURNS TABLE
+DETERMINISTIC
+READS SQL DATA
```

Comment:
Please remove these qualifiers. We don't use them, and we don't determine these properties for table functions in Trino currently.